### PR TITLE
[ISSUE #4029]  apidoc stage2: api-crud

### PR DESCRIPTION
**Pull request by 847850277**
Make sure that:

- [ ] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [ ] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ApiController.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+     * @param state       state.
+     * @param currentPage current page.
+     * @param pageSize    page size.
+     * @return {@linkplain ShenyuAdminResult}
+     */
+    @GetMapping("")
+    public ShenyuAdminResult queryApis(final String apiPath, final Integer state,
+                                       @NotNull final Integer currentPage,
+                                       @NotNull final Integer pageSize) {
+        CommonPager<ApiVO> commonPager = apiService.listByPage(new ApiQuery(apiPath, state, new PageParameter(currentPage, pageSize)));
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ApiController.java:75]: `queryApis` maybe add a `tagId` condition ?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ApiDTO.java [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    /**
+     * create time.
+     */
+    private Date dateCreated;
+
+    /**
+     * update time.
+     */
+    private Date dateUpdated;
+
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ApiDTO.java:119]: lost `tagId`

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/vo/ApiVO.java [hunk 3]
```diff
@@ -0,0 +1,659 @@
...
+    /**
+     * create time.
+     */
+    private Date dateCreated;
+
+    /**
+     * update time.
+     */
+    private Date dateUpdated;
+
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/model/vo/ApiVO.java:116]: add `TagVO` field which can be null.

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ApiController.java (hunk 1 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ApiController.java]: fixed it

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ApiDTO.java (hunk 2 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ApiDTO.java:120]: fixed it

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/vo/ApiVO.java (hunk 3 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/model/vo/ApiVO.java:117]: fixed it

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    @Transactional(rollbackFor = Exception.class)
+    public String delete(final List<String> ids) {
+        // select api id.
+        List<ApiDO> apis = this.apiMapper.selectByIds(ids);
+        if (CollectionUtils.isEmpty(apis)) {
+            return AdminConstants.SYS_API_ID_NOT_EXIST;
+        }
+        // delete apis.
+        this.apiMapper.deleteByIds(ListUtil.map(apis, ApiDO::getId));
+        return StringUtils.EMPTY;
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java:130]: maybe lost `tagRelationMapper.deleteByApiId` ?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    @Override
+    public ApiVO findById(final String id) {
+        return apiMapper.selectByPrimaryKeyGetVO(id);
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java:135]: It is not recommended to directly query `VO` from the database, just query `DO` and then assemble it to `VO`

File: shenyu-admin/src/main/resources/mappers/api-sqlmap.xml [hunk 6]
```diff
@@ -257,4 +287,106 @@
...
+      api.document_md5,
+      api.date_created,
+      api.date_updated,
+      tag.id as tag_id,
+      tag.date_created as tag_date_created,
+      tag.date_updated as tag_date_updated,
+      tag.name as tag_name,
+      tag.tag_desc as tag_desc,
+      tag.parent_tag_id as tag_parent_tag_id,
+      tag.ext as tag_ext
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/resources/mappers/api-sqlmap.xml:385]: `api-sqlmap` only handles `ApiDO` and should not query from `tag_relation` or `tag`, it is recommended to assemble complex `VOs` on the `Service` layer.

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ApiController.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * detail plugin.
+     *
+     * @param id plugin id.
+     * @return {@linkplain ShenyuAdminResult}
+     */
+    @GetMapping("/{id}")
+    @RequiresPermissions("system:api:edit")
+    public ShenyuAdminResult detailApi(@PathVariable("id")
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ApiController.java:89]: Why do query operations require permission?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/mapper/ApiMapper.java [hunk 8]
```diff
@@ -72,4 +79,44 @@ public interface ApiMapper {
...
+    @Override
+    Boolean existed(@Param("id") Serializable id);
+
+    /**
+     * select api by query.
+     *
+     * @param query {@linkplain ApiQuery}
+     * @return {@linkplain List}
+     */
+    List<ApiVO> selectByQuery(ApiQuery query);
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/mapper/ApiMapper.java:98]: change to `List<ApiDO> selectByQuery(ApiQuery query);`

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/mapper/ApiMapper.java [hunk 9]
```diff
@@ -72,4 +79,44 @@ public interface ApiMapper {
...
+     */
+    int deleteByIds(List<String> ids);
+
+    /**
+     * select by primary key get VO.
+     *
+     * @param id primary key
+     * @return object by primary key VO
+     */
+    ApiVO selectByPrimaryKeyGetVO(@Param(value = "id") String id);
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/mapper/ApiMapper.java:121]: remove `ApiVO selectByPrimaryKeyGetVO(@Param(value = "id") String id);`, assemble it in the `Service` layer.

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    @Override
+    public CommonPager<ApiVO> listByPage(final ApiQuery apiQuery) {
+        return PageResultUtils.result(apiQuery.getPageParameter(), () -> apiMapper.selectByQuery(apiQuery));
```
Comment (dragon-zhang) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java:140]: It is not recommended to directly query `VO` from the database, just query `DO` and then assemble it to `VO`

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java (hunk 4 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java:140]: Not required for batch deletion?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java (hunk 5 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java]: fixed it

File: shenyu-admin/src/main/resources/mappers/api-sqlmap.xml (hunk 6 above)
Comment (847850277) [reply shenyu-admin/src/main/resources/mappers/api-sqlmap.xml]: fixed it

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ApiController.java (hunk 7 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ApiController.java:88]: fixed it

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/mapper/ApiMapper.java (hunk 8 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/mapper/ApiMapper.java]: fixed it

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/mapper/ApiMapper.java (hunk 9 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/mapper/ApiMapper.java]: fixed it

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java (hunk 10 above)
Comment (847850277) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java]: fixed it

**Review by dragon-zhang — APPROVED** [review]

**Comment by dragon-zhang** [comment]
This will be merged after CI passed.

**Pull request merged**